### PR TITLE
[Mem2Reg] Tied lexical borrow to initialization.

### PR DESCRIPTION
**Pull request by nate-chandler**
Previously, the lexical borrow scopes that were introduced to replace lexical stack allocations were tied to the uses of the stored value.  That meant that the borrow scope could be both too long and also too short.  It could be too long if there were uses of the value after the dealloc stack and it would be too short if there were not uses of the value while the value was still stored into the stack allocation.

Here, instead, the lexical borrow scopes are tied to the storage of a value into the stack allocation.  A value's lexical borrow scope begins when the storage is initialied with the value; its lexical borrow scope ends when the storage is deinitialized.  That corresponds to the range during which a var in the Swift source has a particular value assigned to it.

Mem2Reg's implementation is split into a few steps:
(1) visiting the instructions in a basic block which contains all uses of an alloc_stack
(2.a) visiting the instructions in each basic block which contains a use of the alloc_stack
(2.b) adding phis
(2.c) using the last stored values as arguments to the new outgoing phi arguments
(2.c) replacing initial uses of the storage with the new incoming phi arguments
And here, (1) amounts to a special case of (2.a).

During (1) and (2.a):
(a) lexical borrow scopes are begun after store instructions for the values that were stored
(b) when possible, lexical borrow scopes are ended before instructions that deinitialize memory
    - destroy_addr
    - store [assign]
    - load [take]
For (1), that is enough to create valid borrow scopes.

For (2), there are two complications:
(a) Borrow scopes that are begun may not be ended (when visiting a single block's instructions).

    For example, when visiting

    ```
    bb1:
      store %instance to [init] %addr
      br bb2
    ```

    a borrow scope is started after the store but cannot be ended.

(b) There may not be enough information available to end borrow scopes when visiting instructions that would typically introduce borrow scopes.

    For example, when visiting

    ```
    bb2:
      %copy = load [copy] %addr
      %instance = load [take] %addr
      br bb3
    ```

    there is not enough information available to end the borrow scope that should be ended before the `load [take]`.

To resolve these issues, both sorts of instructions are tracked.  For (a), in `StackAllocationPromoter::unendedLexicalLifetimes`.  For (b), in `StackAllocationPromoter::unprecedentedDeinitializations`.  Finally, a new step is added:

(2.d) `StackAllocationPromoter::endLexicalLifetimes` does a forward CFG walk starting from the out-edge of each of the blocks which began but did not end lexical lifetimes.  At an out-edge, we do a check regarding unreachables is done, and we may end the borrow scope.  Otherwise, the walk continues to the in-edge of each successor.  At an in-edge, we look for an instruction from (b) (in `unprecedentedDeinitializations`) above.  If one is found, then we end the borrow scope before that instruction.  Otherwise, the walk continues to the out-edge of the block.

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
Together with: https://github.com/apple/swift-source-compat-suite/pull/606
@swift-ci please test source compatibility

**Comment by atrick** [comment]
I didn't expect this because we should never extend the borrow scope beyond the original `dealloc_stack`. Does the algorithm ensure that won't happen independent of this fix?

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@atrick Thank you!  Regarding

> Does the algorithm ensure that won't happen independent of this fix?

No, it didn't.

I've repurposed this PR to fix that bug (keeping all the tests on terminators as well).  Now, the logic for where to end the lifetime in a block which has been identified for lifetime ending is: (1) if there's a new `store` to the address, end the scope there; (2) if the storage is deallocated, end the scope there; (3) otherwise, end the scope at the end of the block.  When deciding where in the block to end the lifetime, there's no longer any consideration of the uses.

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

File: lib/SILOptimizer/Transforms/SILMem2Reg.cpp [hunk 1]
```diff
@@ -68,6 +68,19 @@ struct LiveValues {
...
 
+  /// When iterating through the instructions in a block, whether, when the
+  /// storage stops storing a value, the lexical lifetime for that value should
+  /// be ended at that point in the execution of the pass.
+  ///
+  /// The lifetime should not be ended there if we don't yet have enough
+  /// information to end it.  That occurs when the running values are from a
+  /// load which was not preceeded by a store.  In that case, the lifetime end
+  /// will be added later, when we have enough information (the phi arguments)
+  /// to end it.
```
Comment (atrick) [inline lib/SILOptimizer/Transforms/SILMem2Reg.cpp:79]: It's hard to relate this discussion to the code. I think it's saying: return true if promoteAllocationInBlock has enough information to end the borrow scope. If this returns false, then the scope ending instructions will be emitting during fixBranchesAndUses

File: lib/SILOptimizer/Transforms/SILMem2Reg.cpp [hunk 2]
```diff
@@ -980,15 +980,21 @@ void StackAllocationPromoter::fixBranchesAndUses(BlockSetVector &phiBlocks,
...
+///
+///     In order for a block to propagate the borrowed value so that it might
+///     eventually be end_borrow'd, the block must pass that value on to all its
+///     successors; if any successor does not take that value, then neither that
+///     successor nor any of its own successors could possibly destroy the
+////    value.
+/// (3) The block had a dealloc_stack.
+///
+///     If the allocated space was deallocated in this block, the lifetime of
+///     any value stored in it must end there.
```
Comment (atrick) [inline lib/SILOptimizer/Transforms/SILMem2Reg.cpp:1077]: I think we should clarify the requirements of this algorithm vs the implementation. The lexical scope needs to end at any of
1. dealloc_stack
2. function exit
3. iterated dominance frontier relative to the points where the alloc_stack is initialized

Note that loaded value lifetimes and destroy_value uses should have no effect.

Only case #3 is interesting. We can simplify the problem by noting that it is valid to end the lexical lifetime anywhere between the stack location's deinitialization and its dealloc_stack. 

We should note somewhere that it is illegal for any SIL pass to convert this
```
%v = load [copy] %a
...<stuff>
destroy %v
destroy_addr %a
```

into
```
%v = load [take] %a
...<stuff>
destroy %v
```

Unless the pass proves that the destroy can safely be hoisted above <stuff> according the same logic that lets us shrink borrow scopes.

Then we can reason that all paths that have a lexical borrow must either end in a function exit or destroy_addr (or [take]).

We may still need to be careful about dominance though. Can we handle this case?

```
bb1
  br bb5

bb2:
  store [init]
  cond_br bb3, bb4

bb3:
  destroy_addr

bb4:
  br bb5

bb5:
  unreachable
```

File: lib/SILOptimizer/Transforms/SILMem2Reg.cpp (hunk 1 above)
Comment (nate-chandler) [reply lib/SILOptimizer/Transforms/SILMem2Reg.cpp:79]: How about

```
  /// Returns true if we have enough information to end the lifetime during
  /// promoteAllocationInBlock.
  ///
  /// The lifetime cannot be ended there if we don't yet have enough
  /// information to end it.  That occurs when the running values are from a
  /// load which was not preceeded by a store.  In that case, the lifetime end
  /// will be added later, when we have enough information (e.g., the phi
  /// arguments) to end it.
  bool canEndLexicalLifetimeInPromoteAllocationInBlock(AllocStackInst *asi) {
    return shouldAddLexicalLifetime(asi) && borrow;
  }
```

**Comment by nate-chandler** [comment]
@swift-ci please test

File: lib/SILOptimizer/Transforms/SILMem2Reg.cpp [hunk 3]
```diff
@@ -957,50 +1056,125 @@ void StackAllocationPromoter::fixBranchesAndUses(BlockSetVector &phiBlocks,
...
+  // push (bb3, ::In).  Next, we visit (bb2, ::In).  We see that it _does_ have
+  // an incoming unopened borrow scope, so we close it and stop.  Finally, we
+  // visit (bb3, ::Out).  We see that it too has an incoming unopened borrow
+  // scope so we close it and stop.
+  enum class LiveValuesKind : uint8_t { In, Out };
+
+  using ScopeEndPosition = std::pair<SILBasicBlock *, LiveValuesKind>;
+
+  DAGNodeWorklist<ScopeEndPosition, 16, llvm::SmallSet<ScopeEndPosition, 16>>
+      worklist;
```
Comment (atrick) [inline lib/SILOptimizer/Transforms/SILMem2Reg.cpp:1140]: 1. Can you use PointerIntPair? Then you don't need any special worklist template parameters.

2. Better yet, it seems that the code would be a lot easier to understand if had two worklists, one for each purpose. No need for a `LiveValuesKind` enum

3. Better yet, can you just handle the `LiveValuesKind::In` case before pushing that block onto the regular worklist. Then you just have a normal forward CFG traversal, with the usual worklist. You just have two separate points in the traversal where you handle the In vs. Out case

**Review by atrick — COMMENTED** [review]
Commenting on just the `endLexicalLifetime` algorithm...

**Review by atrick — COMMENTED** [review]
With this extra layer of logic added, I think Mem2Reg as a whole would
be a lot easier to approach if we make it clear that it is only
tracking two symmetric things for each block:

- cross-block initialization points

- cross-block deinitialization points

Then we also don't need to separately track valid storage.

We really only need to track one instruction in each case: the store or the destroy.

It would help to make it clear that the other data types, LiveValues,
and ValueInstructions are transient values only used while processing
a single basic block. They are derived for each basic block either
from the cross-block map or from the block's arguments. Then they can
be updated during block traversal.

> StoreInst *store = nullptr;

I would keep in code that defensively initializes types.

>  // When visiting (bb1, ::Out), we see that bb1 is neither unreachable nor
  // has any successors without live-in values.  So we add (bb2, ::In) to the

In a few places we talk about checking for "any successors". It
implies that the algorithm ends the scopes when there are some
successors that have live-in values and some that don't. That
fundamentally doesn't makes sense. Instead, let's make it clear that
that we end the scope only when all successors have no live-in values,
and that implies all paths lead to unreachable. The implementation
ensures this only happens at a CFG merge, so there will only be one
successor to check.

File: lib/SILOptimizer/Transforms/SILMem2Reg.cpp (hunk 3 above)
Comment (nate-chandler) [reply lib/SILOptimizer/Transforms/SILMem2Reg.cpp]: As discussed, went with 1 because I think it most clearly shows how the walk proceeds at each step.

**Comment by nate-chandler** [comment]
@swift-ci please test

**Review by atrick — APPROVED** [review]
Thanks!

**Comment by nate-chandler** [comment]
@swift-ci please clean test macos platform

**Pull request merged**